### ISSUE TITLE
-999 instead of 401

### ISSUE THREAD

**Issue by grosch** [Alamofire/Alamofire#1159]
When doing a .GET with authentication, things work fine if I give the right user/pwd.  If I give the incorrect user or pwd with Alamofire 3.2.1, response.response from the responseJSON method is getting nil, and the error being passed to the .Failure enum is -999.

My request manager is NOT going out of scope.

**Comment by cnoon** [comment]
Hi @grosch,

These types of questions are best suited for Stack Overflow. We use GitHub for bug reports and feature requests.

Best of luck! 🍻

---

From our [Contribution Guidelines](https://github.com/Alamofire/Alamofire/blob/master/CONTRIBUTING.md)

## Asking Questions

We don't use GitHub as a support forum. For any usage questions that are not specific to the project itself, please ask on [Stack Overflow](https://stackoverflow.com) instead. By doing so, you'll be more likely to quickly solve your problem, and you'll allow anyone else with the same question to find the answer. This also allows maintainers to focus on improving the project for others.

**Comment by tieleman** [comment]
I've seen multiple issues here on Github on the 401 status code (when providing bad credentials for example) that are being closed as support questions or as "works as intended". While I agree that cancelling the request is somewhat defendable when authentication doesn't work, the fact that you cannot access the response body is inconvenient.

The HTTP spec suggests that the user agent should be able to use the response body (https://tools.ietf.org/html/rfc7235#section-3.1):

> If the 401 response contains the same challenge as the prior response, and the user agent has already attempted authentication at least once, then the user agent SHOULD present the enclosed representation to the user, since it usually contains relevant diagnostic information.

Currently Alamofire prevents developers from actually accessing this content unless they insert the `Authorization` header manually instead of using the `authenticate` method (at least, that's one of the workaround I've seen around here). For example, a lot of API's return JSON-encoded error messages to present to the user.

Any thoughts?

**Comment by grosch** [comment]
I wholeheartedly agree with @tieleman.  There's a big difference between "The request was canceled" and "Your credentials were incorrect"

**Comment by yusuftor** [comment]
@tieleman is right and @grosch was right to report this as a bug. I spent about 3 hours today trying to work out why the request returned "cancelled" instead of a 401 error because I want to display the response body to the user. What's the reasoning behind this?

**Comment by tieleman** [comment]
I think the main culprit and source of confusion here is the fact that in the README it lists two separate ways to do HTTP Basic Auth (https://github.com/Alamofire/Alamofire#http-basic-authentication), but it isn't made clear that they will trigger entirely different behaviour:

* One using `authenticate`, which will cancel your request (and thus trigger a failure) without providing any clues as to why this happened and you cannot access the status code or response body.
* One using a manual insertion of the `Authorization` header which will trigger a success (that you can turn into a failure using `validate` if you want) which _does_ give you access to the status code and the response body.

Now, the `authenticate` method is obviously more elegant to use, but using it withholds important information from the developer (and `validate` doesn't work on this as you'd expect). This might just be caused by the fact that `NSURLCredential` triggers this behaviour and Alamofire can do nothing to prevent that, but it should be made clear (and hence I guess this was labelled as a support question) that this difference in behaviour exists.

**Comment by cnoon** [comment]
Okay everyone, apologies for closing this out prematurely. Moved through the issue too quickly. Reopening.

As you have all clearly pointed out, this is certainly worth investigating further to see if there's a better way we can bubble this back to the user. AFN has been working on a solution for when cert or public key pinning evaluation fails, but I'm still a bit weary of such a change. IMO, if we're going to give custom errors for cert or public key pinning failures, we should certainly do the same for other authorization challenge types.

For anyone that feels strongly, I'd encourage them to first ramp up on the efforts from AFN, just so that you have all the proper context. Then, let's discuss the possible options here in this issue.

* [AFN Background](https://github.com/AFNetworking/AFNetworking/issues/3165)
* [AFN Server Trust Error PR](https://github.com/AFNetworking/AFNetworking/pull/3425)

**Comment by tieleman** [comment]
Thanks for reopening. I have reviewed the two linked AFN issues, and while there are some similarities, I do believe that they cater to a different class of problems (cert or key pinning failures). These are typical examples of errors that occur at the secure transport level, while the issue with 401's (and other HTTP status codes) resides at the application level after the server has attempted to process the request.

There's also a slight discrepancy between the described behaviour of Alamofire in the docs and what actually happens when you insert the `authenticate` method. The docs state:

> By default, Alamofire treats any completed request to be successful, regardless of the content of the response. Calling validate before a response handler causes an error to be generated if the response had an unacceptable status code or MIME type.

Using `authenticate` breaks this default (without warning) and renders `validate` unusable. Therefore, it makes sense (to me) to have cert and key failures trigger a failure (because the request is denied before it is even processed on the application server), but have 401's and other faulty status codes trigger a success (as is the default per the docs) and the ability to force that into a failure with `validate`.

In other words: `authenticate` should simply insert credentials into the request but not change the default behaviour. Pinning certificates that fail validation should of course trigger a failure/cancellation.

Does that make sense in this context?

**Comment by cnoon** [comment]
@0xced has resolved this issue in #1164. I'm going to close this issue out for now. Please direct all future comments to #1164.

Cheers. 🍻
